src/server/pkg/collection/collection.go

```diff
@@ -493,77 +479,95 @@ func (c *readonlyCollection) GetBlock(key string, val proto.Message) error {
-	resp, err := c.etcdClient.Get(c.ctx, queryPrefix, etcd.WithPrefix(), etcd.WithLimit(QueryPaginationLimit))
-	if err != nil {
-		return nil, err
-	}
-	return &paginatedIterator{
-		endKey:     endKeyFromPrefix(queryPrefix),
-		resp:       resp,
-		etcdClient: c.etcdClient,
-		ctx:        c.ctx,
-		col:        c,
-	}, nil
+	return c.listF(queryPrefix, &c.limit, order, func(kv *mvccpb.KeyValue) error {
+		if err := proto.Unmarshal(kv.Value, val); err != nil {
+			return err
+		}
+		return f(strings.TrimPrefix(string(kv.Key), queryPrefix))
+	})
 }
 
-// List returns an iterator that can be used to iterate over the collection.
-// The objects are sorted by revision time in descending order, i.e. newer
-// objects are returned first.
-func (c *readonlyCollection) List() (Iterator, error) {
-	resp, err := c.etcdClient.Get(c.ctx, c.prefix, etcd.WithPrefix(), etcd.WithSort(etcd.SortByModRevision, etcd.SortDescend))
-	if err != nil {
-		return nil, err
+// ListF returns objects sorted by CreateRevision, i.e. the order in which they
+// were created. f will be called with each key, val will contain the
+// corresponding value. Val is not an argument to f because that would require
+// f to perform a cast before it could be used.
+// You can break out of iteration by returning errutil.ErrBreak.
+func (c *readonlyCollection) ListF(order Order, val proto.Message, f func(string) error) error {
+	if err := watch.CheckType(c.template, val); err != nil {
+		return err
 	}
-	return &iterator{
-		resp: resp,
-		col:  c,
-	}, nil
+	return c.listF(c.prefix, &c.limit, order, func(kv *mvccpb.KeyValue) error {
+		if err := proto.Unmarshal(kv.Value, val); err != nil {
+			return err
+		}
+		return f(strings.TrimPrefix(string(kv.Key), c.prefix))
+	})
 }
 
-// ListPaginated returns an iterator that can be used to iterate over the collection.
-// The objects are not sorted, but are paginated.
-func (c *readonlyCollection) ListPaginated() (Iterator, error) {
-	resp, err := c.etcdClient.Get(c.ctx, c.prefix, etcd.WithPrefix(), etcd.WithLimit(QueryPaginationLimit))
-	if err != nil {
-		return nil, err
+func (c *readonlyCollection) listF(prefix string, limitPtr *int64, order Order, f func(*mvccpb.KeyValue) error) error {
+	rev := int64(-1)
+	keysInRev := make(map[string]bool)
+	for {
+		sort := etcd.WithSort(etcd.SortByCreateRevision, etcd.SortDescend)
+		if order == Ascend {
+			sort = etcd.WithSort(etcd.SortByCreateRevision, etcd.SortAscend)
+		}
+		opts := []etcd.OpOption{etcd.WithPrefix(), sort}
+		if rev != -1 {
+			if order == Descend {
+				opts = append(opts, etcd.WithMaxCreateRev(rev))
+			} else {
+				opts = append(opts, etcd.WithMinCreateRev(rev))
+			}
+		}
+		var resp *etcd.GetResponse
+		var limit int64
+		for {
+			limit = atomic.LoadInt64(limitPtr)
+			var err error
+			resp, err = c.etcdClient.Get(c.ctx, prefix, append(opts, etcd.WithLimit(limit))...)
+			if err != nil {
+				if strings.Contains(err.Error(), "received message larger than max") {
```

Comment:
It might be better to check against an error code rather than the message because I have seen this message be different between grpc versions. It looks like the error code for this is ResourceExhausted.